fw/cache.c

```diff
@@ -2824,6 +2857,20 @@ tfw_cache_build_resp(TfwHttpReq *req, TfwCacheEntry *ce, long lifetime,
 	}
 	resp->content_length = ce->body_len;
 
+	if (has_trailers) {
```

Comment:
As I see most of this code is necessary only because we don't use pool for making trailer headers. I am right? @krizhanovsky how do you think may be we should rework pool related functions and use them for making trailers?

Reply:
There are two paths to generate responses.
1. backend --> client, this path uses a pool
2. cache --> client, this path has no pool, but just bundles all headers into a skb list

both paths have different logic and functions to construct responses, and both assume no trailers, moreover, in the prepare_xmit hook, it also assumes such layout of the skb list.
I checked the code of both paths before, and I think it's very very complicated to do refactoring (too many details are involved), I think it is not worthing to pay such a big cost for a relatively trial RFC fix. To be honest, even regarding my current hacking way, it took me a week.

Reply:
@EvgeniiMekhanik @krizhanovsky @const-t 
If I understand correctly after checking the code `__tfw_http_msg_expand_from_pool`, the so-called "pool" is just used to linearize the regular headers and the response body. That is, if use pool, the memory contains headers and body is linear, otherwise, it's scattered into multiple skbs. But that's okay for trailers, and more precisely, it's better for trailers:
1. the trailers are encoded separately from regular headers and body
2. the trailers' skbs have different flags which must not be mixed with regular headers and body, in the latter case, small headers list and body may share one skb, if mixed with trailers too, that's complicated.

So, IMO, KISS (keep it simple and stupid) is better.